cache/class-vary-cache.php

```diff
@@ -0,0 +1,292 @@
+	 *
+	 * @since   1.0.0
+	 * @access  public
+	 *
+	 * @return array  user's group-value pairs
+	 */
+	public static function get_user_groups() {
+		self::parse_group_cookie();
+		return self::$groups;
+	}
+
+	/**
+	 * Sets the context of the the group segmentation to be encrypted or not.
+	 *
+	 * @since   1.0.0
+	 * @access  public
+	 *
+	 * @param bool $value true for encrypted requests.
+	 */
+	public static function set_encryption( $value = true ) {
+		static::$encryption_enabled = $value;
+	}
+
+	/**
+	 * Returns the encryption flag
+	 *
+	 * @since   1.0.0
+	 * @access  public
+	 *
+	 * @return bool true if encryption is set for this request
+	 */
+	public static function is_encryption_enabled() {
+		return static::$encryption_enabled;
+	}
+
+	/**
+	 * Encrypts a string using the auth credentials for the site.
+	 *
+	 * @since   1.0.0
+	 * @access  private
+	 *
+	 * @param string $value cookie text value.
+	 * @throws string If credentials ENV Variables aren't defined.
+	 * @return string encrypted version of string
+	 */
+	private static function encrypt_cookie_value( $value ) {
+		// Validate that we have the secret values.
+		if ( ! defined( 'VIP_GO_AUTH_COOKIE_KEY' ) || ! defined( 'VIP_GO_AUTH_COOKIE_IV' ) ) {
+			// TODO: check that values are not empty.
+			trigger_error( 'Secrets not defined for encrypted vary cookies', E_USER_WARNING );
+			return;
+		}
+
+		$client_key = constant( 'VIP_GO_AUTH_COOKIE_KEY' );
+		$client_iv = constant( 'VIP_GO_AUTH_COOKIE_IV' );
+		$cookie_value = random_bytes( 32 ) . '|' . $value . '|' . ( time() + self::$cookie_expiry );
+		$cipher_cookie = openssl_encrypt( $cookie_value, 'aes-128-cbc', $client_key, 0, $client_iv );
+
+		return $cipher_cookie;
+	}
+
+	/**
+	 * Decrypts a string using the auth credentials for the site.
+	 *
+	 * @since   1.0.0
+	 * @access  private
+	 *
+	 * @param string $cookie_value the encrypted string.
+	 * @throws string If credentials ENV Variables aren't defined.
+	 * @return string decrypted version of string
+	 */
+	private static function decrypt_cookie_value( $cookie_value ) {
+		// Validate that we have the secret values.
+		if ( ! defined( 'VIP_GO_AUTH_COOKIE_KEY' ) || ! defined( 'VIP_GO_AUTH_COOKIE_IV' ) ) {
+			// TODO: check that values are not empty.
+			trigger_error( 'Secrets not defined for encrypted vary cookies', E_USER_WARNING );
+			return;
+		}
+
+		$client_key = constant( 'VIP_GO_AUTH_COOKIE_KEY' );
+		$client_iv = constant( 'VIP_GO_AUTH_COOKIE_IV' );
+		$cipher_cookie = openssl_decrypt( $cookie_value, 'aes-128-cbc', $client_key, 0, $client_iv );
+		$cookie_array = explode( '|', $cipher_cookie );
+		// Parse out the group payload (2nd item).
+		if ( count( $cookie_array ) < 2 ) {
+			return null;
+		}
+		return $cookie_array [1];
+	}
+
+	/**
+	 * Parses the text cookie into the local groups array of key-values.
+	 *
+	 * @since   1.0.0
+	 * @access  private
+	 */
+	private static function parse_group_cookie() {
```

Comment:
This may end up being called several times during a pageload (e.g. if I call `get_user_groups` and `is_user_in_group` on the same pageload). We should see if that's still necessary. If it is, try to reduce the overhead. Can be tracked in a separate issue + PR.

Reply:
Okay, I'll track that in an issue